src/log.c

```diff
@@ -36,71 +36,294 @@
+					 prov_name);
+
+	if (found) {
+		if (provider_list.negated)
+			log_mask &= ~EXPAND(position, PROV_OFFSET);
 		else
-			fi_log_level = (int)ret;
+			log_mask |= EXPAND(position, PROV_OFFSET);
+
+		item = container_of(found, struct provider_parameter, entry);
+		free(item->name);
+		free(item);
+	}
+}
+
+/*
+ * Negatable describes whether the function should check for the caret symbol
+ * (^) at the beginning of the given environment variable. Tokenize string input
+ * and run given action function on each token found.
+ */
+static bool find_options(bool negatable, char *input, found_func_t *action)
+{
+	bool negated = false;
+	char *save_ptr = NULL;
+	char *token = NULL;
+
+	if (!input)
+		return negated;
+
+	if (negatable) {
+		if (*input == '^') {
+			input++;
+			negated = true;
+		}
+	}
+
+	for (token = strtok_r(input, ",", &save_ptr); token;
+	     token = strtok_r(NULL, ",", &save_ptr)) {
+		action(token);
+	}
+
+	return negated;
+}
+
+static void handle_log_option(char *name)
+{
+	bool found = false;
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(log_levels); i++) {
+		if (log_levels[i] && !strcasecmp(log_levels[i], name)) {
+			found = true;
+			break;
+		}
+	}
+
+	if (found) {
+		log_mask |= i;
+	} else {
+		fprintf(stderr,
+			"%s: invalid option \"%s\" for env FI_LOG_LEVEL.\n",
+			PACKAGE, name);
+	}
+}
+
+static void handle_subsys_option(char *name)
+{
+	bool found = false;
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(subsystems); i++) {
+		if (subsystems[i] && !strcasecmp(subsystems[i], name)) {
+			found = true;
+			break;
+		}
+	}
+
+	if (found) {
+		log_mask |= EXPAND(i, SUBSYS_OFFSET);
+	} else {
+		fprintf(stderr,
+			"%s: invalid option \"%s\" for env FI_LOG_SUBSYSTEMS.\n",
+			PACKAGE, name);
+	}
+}
+
+static void handle_provider_option(char *name)
+{
+	struct provider_parameter *param = calloc(1, sizeof(*param));
+
+	if (!param) {
+		fprintf(stderr, "%s: %s", __func__, strerror(errno));
+		return;
 	}
+
+	/*
+	 * Insert into provider_list for lookup during registration.
+	 */
+	param->name = strdup(name);
+	slist_insert_tail(&param->entry, &provider_list.names);
+}
+
+static void get_log_level(void)
+{
+	char *requested;
+
+	requested = alloc_env("FI_LOG_LEVEL");
+	if (!requested)
+		return;
+
+	find_options(false, requested, handle_log_option);
+	free(requested);
+}
+
+static void get_providers(void)
+{
+	char *requested;
+
+	requested = alloc_env("FI_LOG_PROV");
+	if (!requested)
+		return;
+
+	provider_list.negated = find_options(true, requested,
+					     handle_provider_option);
+
+	if (!provider_list.negated)
+		log_mask &= ~PROV_MASK;
+
+	free(requested);
 }
 
-void fi_warn_impl(const char *prov, const char *fmt, ...)
+static void get_subsystems(void)
 {
+	char *requested;
+
+	requested = alloc_env("FI_LOG_SUBSYSTEMS");
+	if (!requested)
+		return;
+
+	log_mask &= ~SUBSYS_MASK;
+
+	/*
+	 * find_options returns whether the requested options were negated
+	 */
+	if (find_options(true, requested, handle_subsys_option))
```

Comment:
Needing a comment like this indicates that the use of the function isn't clear.  It would be better to pass negated as a parameter.